src/org/rascalmpl/values/iterators/DescendantReader.java

```diff
@@ -105,6 +116,19 @@ else if(type.isTuple()) {
 			spine.push(v);
 		}
 	}
+
+	private void pushAmb(ITree amb) {
+		if (debug) System.err.println("pushAmb: " + amb);
+		spine.push(amb);
+		for (IValue alt : TreeAdapter.getAlternatives(amb)) {
```

Comment:
why not hash on the entire amb cluster? then there are 50% (or more) fewer nodes to cache.